pontoon/base/views.py

```diff
@@ -348,6 +348,87 @@ def get_translation_history(request):
     return JsonResponse(payload, safe=False)
 
 
+@utils.require_AJAX
+@login_required(redirect_field_name='', login_url='/403')
+@transaction.atomic
+def approve_translation(request):
+    """Approve given translation."""
+    try:
+        t = request.POST['translation']
+        paths = request.POST.getlist('paths[]')
+    except MultiValueDictKeyError as e:
+        return HttpResponseBadRequest('Bad Request: {error}'.format(error=e))
+
+    translation = get_object_or_404(Translation, pk=t)
+    project = translation.entity.resource.project
+    locale = translation.locale
+    user = request.user
+
+    # Read-only translations cannot be approved
+    if utils.readonly_exists(project, locale):
+        return HttpResponseForbidden(
+            "Forbidden: This translation is in read-only mode"
+        )
+
+    if translation.approved:
+        return HttpResponseForbidden(
+            "Forbidden: This translation is already approved"
+        )
+
+    # Only privileged users can approve translations
+    if not user.can_translate(locale, project):
+        return HttpResponseForbidden(
+            "Forbidden: You don't have permission to approve this translation."
+        )
+
+    # Check for errors.
+    # Checks are disabled for the tutorial.
+    use_checks = project.slug != 'tutorial'
+
+    if use_checks and translation.errors.exists():
+        return JsonResponse({
+            'failedChecks': translation.errors.all(),
+        })
+
+    now = timezone.now()
+
+    # Reject previously approved translations.
+    translations = Translation.objects.filter(
+        entity=translation.entity,
+        locale=locale,
+        plural_form=translation.plural_form,
+    )
+    translations.filter(approved=True).update(
+        active=False,
+        approved=False,
+        rejected=True,
+        rejected_user=user,
+        rejected_date=now,
+    )
+
+    # Make sure there is no other active translation.
+    translations.filter(active=True).update(active=False)
+
+    # Update the translation status and save it.
+    translation.active = True
+    translation.approved = True
+    translation.approved_user = user
+    translation.approved_date = now
+    translation.unapproved_user = None
+    translation.unapproved_date = None
+    translation.fuzzy = False
+    translation.rejected = False
+    translation.rejected_user = None
+    translation.rejected_date = None
+
+    translation.save()
```

Comment:
There are a few things here, inconsistent with the related "review" views (unapprove, reject, unreject, delete):

1. `translation.entity.reset_active_translation()` should take care of the `active` flag instead of mangling with it manually.

2. You should replace lines 393 - 424 with a call to `translation.approve()` model instance method, which you'll need to create. The `# Reject previously approved translations.` block is unnecessary - `Translation.save()` takes care of that.

3. And with those two changes in place, I suspect you should be able to lose `user = request.user` (given that we don't use it in other related methods).